File: apps/web/src/features/workspace/wawCryptoContext.ts
/** Closed, non-secret context records for the fixed WAW v1 crypto profile. */
export const WAW_PROTOCOL_ID = 'agentbox-waw/v1'
export const U64_MAX = 18_446_744_073_709_551_615n

export class WAWCryptoError extends Error {
  constructor() {
    super('STREAM_CRYPTO_FAILURE')
    this.name = 'WAWCryptoError'
  }
}

export interface AdmissionTuple {
  readonly attachment_id: string
  readonly workspace_id: string
  readonly project_id: string
  readonly agent_type: 'claude' | 'codex'
  readonly runtime_host_installation_id: string
  readonly runtime_host_installation_revision: string
  readonly auth_epoch: string
  readonly api_authority_epoch: string
  readonly lease_number: string
  readonly generation: string
  readonly binding_revision: string
  readonly mode: 'writer'
  readonly binding_digest: string
}

export interface HandshakeContext extends Omit<AdmissionTuple, 'mode'> {
  readonly runtime_epoch: string
  readonly protocol_id: typeof WAW_PROTOCOL_ID
  readonly crypto_envelope_version: 1
}

export const ADMISSION_KEYS = Object.freeze([
  'attachment_id',
  'workspace_id',
  'project_id',
  'agent_type',
  'runtime_host_installation_id',
  'runtime_host_installation_revision',
  'auth_epoch',
  'api_authority_epoch',
  'lease_number',
  'generation',
  'binding_revision',
  'mode',
  'binding_digest',
] as const)
export const CONTEXT_KEYS = Object.freeze([
  ...ADMISSION_KEYS.filter((key) => key !== 'mode'),
  'runtime_epoch',
  'protocol_id',
  'crypto_envelope_version',
] as const)

/** Accept data properties only; getters, prototypes and hidden keys are not wire data. */
export function exactRecord(
  value: unknown,
  keys: readonly string[],
): Record<string, unknown> {
  if (
    !value ||
    typeof value !== 'object' ||
    ![Object.prototype, null].includes(Object.getPrototypeOf(value))
  )
    throw new WAWCryptoError()
  const own = Reflect.ownKeys(value)
  if (
    own.length !== keys.length ||
    own.some((key) => typeof key !== 'string' || !keys.includes(key))
  )
    throw new WAWCryptoError()
  const result: Record<string, unknown> = {}
  for (const key of keys) {
    const descriptor = Object.getOwnPropertyDescriptor(value, key)
    if (!descriptor || !descriptor.enumerable || !('value' in descriptor))
      throw new WAWCryptoError()
    result[key] = descriptor.value
  }
  return result
}

export function validateU64(value: unknown): string {
  if (
    typeof value !== 'string' ||
    !/^[1-9][0-9]{0,19}$/.test(value) ||
    BigInt(value) > U64_MAX
  )
    throw new WAWCryptoError()
  return value
}

export function validateHex32(value: unknown): string {
  if (typeof value !== 'string' || !/^[a-f0-9]{64}$/.test(value))
    throw new WAWCryptoError()
  return value
}

function validateMembers(record: Record<string, unknown>): void {
  for (const [key, prefix] of [
    ['attachment_id', 'att'],
    ['workspace_id', 'aws'],
    ['project_id', 'prj'],
    ['runtime_host_installation_id', 'wri'],
  ]) {
    const value = record[key]
    if (
      typeof value !== 'string' ||
      !new RegExp(`^${prefix}_[a-f0-9]{32}$`).test(value)
    )
      throw new WAWCryptoError()
  }
  if (record.agent_type !== 'claude' && record.agent_type !== 'codex')
    throw new WAWCryptoError()
  for (const key of [
    'runtime_host_installation_revision',
    'auth_epoch',
    'api_authority_epoch',
    'lease_number',
    'generation',
    'binding_revision',
  ])
    validateU64(record[key])
  validateHex32(record.binding_digest)
}

export function validateAdmission(value: unknown): AdmissionTuple {
  const record = exactRecord(value, ADMISSION_KEYS)
  validateMembers(record)
  if (record.mode !== 'writer') throw new WAWCryptoError()
  return Object.freeze(record) as unknown as AdmissionTuple
}

export function deriveContext(
  admission: unknown,
  runtimeEpoch: unknown,
): HandshakeContext {
  const validated = validateAdmission(admission)
  const record: Record<string, unknown> = {}
  for (const key of ADMISSION_KEYS)
    if (key !== 'mode') record[key] = validated[key]
  return validateContext({
    ...record,
    runtime_epoch: validateU64(runtimeEpoch),
    protocol_id: WAW_PROTOCOL_ID,
    crypto_envelope_version: 1,
  })
}

export function validateContext(value: unknown): HandshakeContext {
  const record = exactRecord(value, CONTEXT_KEYS)
  validateMembers(record)
  validateU64(record.runtime_epoch)
  if (
    record.protocol_id !== WAW_PROTOCOL_ID ||
    record.crypto_envelope_version !== 1
  )
    throw new WAWCryptoError()
  return Object.freeze(record) as unknown as HandshakeContext
}

/** RFC 8785 equivalent for this closed ASCII string / literal integer schema. */
export function canonicalContextBytes(value: unknown): Uint8Array {
  const context = validateContext(value)
  const canonical = Object.fromEntries(
    Object.entries(context).sort(([a], [b]) => (a < b ? -1 : a > b ? 1 : 0)),
  )
  return new TextEncoder().encode(JSON.stringify(canonical))
}

File: packages/agentbox-protocol/src/agentbox_protocol/waw_crypto_context.py
"""Exact non-secret WAW application-crypto context and canonical prologue.

Validation does not authenticate admission metadata or supply a trust provider.
Callers must supply their already-bound admission and Runtime epoch separately.
"""

from __future__ import annotations

import re

import rfc8785

PROTOCOL_ID = "agentbox-waw/v1"
CRYPTO_ENVELOPE_VERSION = 1
MAX_U64 = 2**64 - 1
ADMISSION_KEYS = frozenset(
    {
        "attachment_id",
        "workspace_id",
        "project_id",
        "agent_type",
        "runtime_host_installation_id",
        "runtime_host_installation_revision",
        "auth_epoch",
        "api_authority_epoch",
        "lease_number",
        "generation",
        "binding_revision",
        "mode",
        "binding_digest",
    }
)
CONTEXT_KEYS = (ADMISSION_KEYS - {"mode"}) | {
    "protocol_id",
    "crypto_envelope_version",
    "runtime_epoch",
}
U64_KEYS = frozenset(
    {
        "runtime_host_installation_revision",
        "auth_epoch",
        "api_authority_epoch",
        "lease_number",
        "generation",
        "binding_revision",
    }
)
_IDS = {
    "attachment_id": "att",
    "workspace_id": "aws",
    "project_id": "prj",
    "runtime_host_installation_id": "wri",
}
_DECIMAL = re.compile(r"[1-9][0-9]{0,19}\Z")
_HEX32 = re.compile(r"[a-f0-9]{64}\Z")


class WAWContextError(ValueError):
    """A closed application context is invalid; values are never echoed."""


def validate_u64(value: object) -> str:
    """Validate one positive canonical uint64 JSON string without coercion."""
    if type(value) is not str or _DECIMAL.fullmatch(value) is None or int(value) > MAX_U64:
        raise WAWContextError("invalid positive uint64 string")
    return value


def validate_hex32(value: object) -> str:
    if type(value) is not str or _HEX32.fullmatch(value) is None:
        raise WAWContextError("invalid lowercase 32-byte hexadecimal string")
    return value


def validate_admission(value: object) -> dict[str, str]:
    """Validate all 13 members and return a detached exact scalar dictionary."""
    if type(value) is not dict or value.keys() != ADMISSION_KEYS:
        raise WAWContextError("invalid AdmissionTuple fields")
    result: dict[str, str] = {}
    for key in ADMISSION_KEYS:
        item = value[key]
        if type(item) is not str:
            raise WAWContextError("invalid AdmissionTuple scalar type")
        if key in U64_KEYS:
            validate_u64(item)
        elif key in _IDS:
            if re.fullmatch(_IDS[key] + r"_[a-f0-9]{32}", item) is None:
                raise WAWContextError("invalid AdmissionTuple identifier")
        elif key == "binding_digest":
            validate_hex32(item)
        elif key == "mode" and item != "writer":
            raise WAWContextError("invalid AdmissionTuple mode")
        elif key == "agent_type" and item not in {"claude", "codex"}:
            raise WAWContextError("invalid AdmissionTuple agent type")
        result[key] = item
    return result


def derive_context(admission: object, runtime_epoch: object) -> dict[str, str | int]:
    """Omit validated writer mode and bind the already-known Runtime epoch."""
    bound = validate_admission(admission)
    result: dict[str, str | int] = {key: item for key, item in bound.items() if key != "mode"}
    result.update(
        protocol_id=PROTOCOL_ID,
        crypto_envelope_version=CRYPTO_ENVELOPE_VERSION,
        runtime_epoch=validate_u64(runtime_epoch),
    )
    return result


def validate_context(value: object) -> dict[str, str | int]:
    """Validate exactly 15 context members; no missing wire field is defaulted."""
    if type(value) is not dict or value.keys() != CONTEXT_KEYS:
        raise WAWContextError("invalid HandshakeContext fields")
    if (
        type(value["protocol_id"]) is not str
        or value["protocol_id"] != PROTOCOL_ID
        or type(value["crypto_envelope_version"]) is not int
        or value["crypto_envelope_version"] != CRYPTO_ENVELOPE_VERSION
    ):
        raise WAWContextError("unsupported application crypto profile")
    admission = {key: value[key] for key in ADMISSION_KEYS if key != "mode"}
    # Mode is not a member of C; this reconstructs its fixed derivation domain.
    admission["mode"] = "writer"
    return derive_context(admission, value["runtime_epoch"])


def canonical_context_bytes(value: object) -> bytes:
    """Complete RFC 8785 UTF-8 C bytes used unchanged as the Noise prologue."""
    return rfc8785.dumps(validate_context(value))


__all__ = [
    "ADMISSION_KEYS",
    "CONTEXT_KEYS",
    "CRYPTO_ENVELOPE_VERSION",
    "MAX_U64",
    "PROTOCOL_ID",
    "U64_KEYS",
    "WAWContextError",
    "canonical_context_bytes",
    "derive_context",
    "validate_admission",
    "validate_context",
    "validate_hex32",
    "validate_u64",
]
